### PR TITLE
[ADD] Noise-free quadratic test problem

### PR DESCRIPTION
**Pull request by ltatzel**

### CODE REVIEW

File: deepobs/pytorch/testproblems/noise_free_quadratic.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
+# -*- coding: utf-8 -*-
+"""TODO Description of the testproblem."""
```
Comment (f-dangel) [inline deepobs/pytorch/testproblems/noise_free_quadratic.py]: Resolve the TODO

File: deepobs/pytorch/testproblems/noise_free_quadratic.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    """
+
+    def __init__(
+        self, batch_size, dim, train_size=128, valid_size=128, test_size=128,
+    ):
+        self._dim = dim
+
+        # Check batch size
+        assert batch_size <= min(
+            [train_size, valid_size, test_size]
```
Comment (f-dangel) [inline deepobs/pytorch/testproblems/noise_free_quadratic.py]: Remove `[ ]` and the code will still work.

File: deepobs/pytorch/testproblems/noise_free_quadratic.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+        ), "Batch size exceeds size of training/validation/test set"
+
+        self._train_size = train_size
+        self._valid_size = valid_size
+        self._test_size = test_size
+
+        # This attribute is needed by _make_train_eval_dataloader
+        self._train_eval_size = self._train_size
+
+        super(data_noise_free_quadratic, self).__init__(batch_size)
```
Comment (f-dangel) [inline deepobs/pytorch/testproblems/noise_free_quadratic.py]: `super()` should be sufficient.

File: deepobs/pytorch/testproblems/noise_free_quadratic.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    def _make_test_dataloader(self):
+        """Creates the test data loader."""
+
+        # Test data
+        test_data = torch.rand(self._test_size, self._dim)
+        test_labels = torch.zeros(self._test_size, self._dim)
+        test_dataset = data.TensorDataset(test_data, test_labels)
+        test_loader = self._make_dataloader(test_dataset)
+
+        return test_loader
```
Comment (f-dangel) [inline deepobs/pytorch/testproblems/noise_free_quadratic.py:70]: Question: Is it necessary to set random seeds here, or does this happen outside these function in DeepOBS?

File: deepobs/pytorch/testproblems/noise_free_quadratic.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+
+# Some helper functions
+def set_weights(linear_layer, weights, req_grad):
+    """Set weights in linear layer and choose if these parameters are
+    trainable.
+    """
+    w = linear_layer.weight
+    if weights.shape == w.data.shape:
```
Comment (f-dangel) [inline deepobs/pytorch/testproblems/noise_free_quadratic.py]: remove `.data`

File: deepobs/pytorch/testproblems/noise_free_quadratic.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+    else:
+        raise ValueError("Weights don't have the right shape")
+
+
+def set_biases(linear_layer, biases, req_grad):
+    """Set biases in linear layer and choose if these parameters are
+    trainable.
+    """
+    b = linear_layer.bias
+    if biases.shape == b.data.shape:
```
Comment (f-dangel) [inline deepobs/pytorch/testproblems/noise_free_quadratic.py]: remove `.data`

File: deepobs/pytorch/testproblems/noise_free_quadratic.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        test_data = torch.rand(self._test_size, self._dim)
+        test_labels = torch.zeros(self._test_size, self._dim)
+        test_dataset = data.TensorDataset(test_data, test_labels)
+        test_loader = self._make_dataloader(test_dataset)
+
+        return test_loader
+
+
+# Some helper functions
+def set_weights(linear_layer, weights, req_grad):
```
Comment (f-dangel) [inline deepobs/pytorch/testproblems/noise_free_quadratic.py]: Refactor `set_weights` and `set_biases` into a single function 
```
def set_param(linear_layer, param, param_str, req_grad):
    p = getattr(linear_layer, param_str)

    if param.shape != p.shape:
        raise ValueError("Weights don't have the right shape")     

    p.data = param
    p.requires_grad = req_grad        
```

File: deepobs/pytorch/testproblems/noise_free_quadratic.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+# The network
+def get_noise_free_quadratic_net(H, theta):
+    """Build the network for the noise-free quadratic
+
+    The network is based on the Hessian ``H`` and the vector ``theta``. It
+    is designed such that the MSE loss of the network (which is parameterized
+    by ``theta``) is ``theta.T @ H @ theta`` for arbitrary inputs with labels
+    that are zero.
+    """
+
```
Comment (f-dangel) [inline deepobs/pytorch/testproblems/noise_free_quadratic.py]: remove blank line

File: deepobs/pytorch/testproblems/noise_free_quadratic.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+    The problem (determined by the Hessian and initial network parameters) is
+    defined in the constructor. It is a quadratic problem of the form
+    ``theta.T @ H @ theta``, where ``H`` is the Hessian and ``theta``
+    corresponds to the trainable parameters of the network. They are initially
+    set to ``theta_init``.
+    """
+
+    def __init__(
+        self, batch_size, weight_decay=None,
+    ):
```
Comment (f-dangel) [inline deepobs/pytorch/testproblems/noise_free_quadratic.py]: This should fit in one line

File: deepobs/pytorch/testproblems/noise_free_quadratic.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+        """Here, the quadratic problem is defined. Note that the batch size
+        is arbitrary: since the problem is noise-free, the batch size has no
+        impact on the resulting loss.
+        """
+
```
Comment (f-dangel) [inline deepobs/pytorch/testproblems/noise_free_quadratic.py]: Remove blank line

File: deepobs/pytorch/testproblems/noise_free_quadratic.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+        super(noise_free_quadratic, self).__init__(batch_size, weight_decay)
```
Comment (f-dangel) [inline deepobs/pytorch/testproblems/noise_free_quadratic.py]: `super()`

File: deepobs/pytorch/testproblems/noise_free_quadratic.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+
+        # Check problem
+        self.check_problem()
+        self._dim = self._H.shape[0]
+
+    def check_problem(self):
+        """Make sure that the attributes ``self._H`` and ``self._theta_init``
+        "match" (dimensions) and that the Hessian is symmetric pos. definite.
+        """
+
```
Comment (f-dangel) [inline deepobs/pytorch/testproblems/noise_free_quadratic.py]: Remove blank line

File: deepobs/pytorch/testproblems/noise_free_quadratic.py [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+        # Check symmetric positive definite
+        assert torch.allclose(H.T, H), "Hessian has to be symmetric"
+        H_eigvals, _ = torch.symeig(H, eigenvectors=False)
+        assert torch.all(H_eigvals > 0), "Hessian has to be positive definite"
+
+    def set_up(self):
+        """Initialize the global attributes ``net``, ``data`` and
+        ``loss_function``.
+        """
+
```
Comment (f-dangel) [inline deepobs/pytorch/testproblems/noise_free_quadratic.py]: remove blank line

File: deepobs/pytorch/testproblems/noise_free_quadratic.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+    def get_batch_loss_and_accuracy_func(
+        self, reduction="mean", add_regularization_if_available=True
+    ):
+        """The original method from the base class doesn't work here
+        (especially for the accuracy), so we overwrite it. It is basically a
+        copy of the original method, but we set the accuracy to zero instead
+        of trying to computing it. Note that the accuracy does't make sense
+        as a metric for our particular problem.
+        """
+
```
Comment (f-dangel) [inline deepobs/pytorch/testproblems/noise_free_quadratic.py]: remove blank line

File: deepobs/pytorch/testproblems/noise_free_quadratic.py [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+        def forward_func():
+
+            # Evaluate loss: In evaluation phase no gradient is needed
+            if self.phase in ["train_eval", "test", "valid"]:
+                with torch.no_grad():
+                    outputs = self.net(inputs)
+                    loss = loss_function(outputs, labels)
+            else:
+                outputs = self.net(inputs)
+                loss = loss_function(outputs, labels)
```
Comment (f-dangel) [inline deepobs/pytorch/testproblems/noise_free_quadratic.py]: reduce duplication:
```
with torch.no_grad() if self.phase in [...] else nullcontext():
      outputs = self.net(inputs)
      loss = loss_function(outputs, labels)
```

File: tests/pytorch/testproblems/test_noise_free_quadratic.py [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+SEEDS = [0, 1, 42]
+IDS_SEEDS = [f"seed_value={seed}" for seed in SEEDS]
+
+
+@pytest.mark.parametrize("seed", SEEDS, ids=IDS_SEEDS)
+@pytest.mark.parametrize("dim", DIMS, ids=IDS_DIMS)
+def test_func(seed, dim):
+
+    # Initialize testproblem
+    nf_quadratic = noise_free_quadratic(batch_size=8)
```
Comment (f-dangel) [inline tests/pytorch/testproblems/test_noise_free_quadratic.py:29]: Does this generate any random numbers? If so, it would be better to fix the seed before.

File: tests/pytorch/testproblems/test_noise_free_quadratic.py [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+    # Create random symmetric pos. definite Hessian and `theta_init`
+    torch.manual_seed(seed)
+
+    theta_init = torch.rand(dim)
+    R = torch.rand(dim, dim)
+    H = R @ R.T + 0.01 * torch.diag(torch.ones(dim))
```
Comment (f-dangel) [inline tests/pytorch/testproblems/test_noise_free_quadratic.py]: the second term is a `torch.eye(dim)`

File: tests/pytorch/testproblems/test_noise_free_quadratic.py [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+    for batch_idx in range(NOF_BATCHES):
+
+        # Get some data (the inputs shouldn't affect the loss at all)
+        input, labels = list(train_loader)[batch_idx]
+        input = input.to(device)
+        labels = labels.to(device)
+
+        # Compare the model's loss with the manually computed loss
+        loss_model = loss_function(net(input), labels)
```
Comment (f-dangel) [inline tests/pytorch/testproblems/test_noise_free_quadratic.py:58]: Ideally, one should call `nf_quadratic.get_batch_loss_and_accuracy_func()()` here for the test to be sensitive to changes in the latter. For that one would need access to `input, labels`. Not sure how to do that.

File: tests/pytorch/testproblems/test_noise_free_quadratic.py [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+
+        assert torch.allclose(
+            loss_model, loss_manually
+        ), "The model's loss and the manually computed quadratic loss deviate."
+
+
+if __name__ == "__main__":
+
+    # For debugging
+    test_func(seed=0, dim=10)
```
Comment (f-dangel) [inline tests/pytorch/testproblems/test_noise_free_quadratic.py]: Delete?

File: deepobs/pytorch/testproblems/noise_free_quadratic.py (hunk 4 above)
Comment (ltatzel) [reply deepobs/pytorch/testproblems/noise_free_quadratic.py:70]: I'm not sure if DeepOBS takes care of this. But in this case, since the problem is noise-free (inputs are multiplied by zero in the first layer of the network), it doesn't matter, does it?

File: tests/pytorch/testproblems/test_noise_free_quadratic.py (hunk 18 above)
Comment (ltatzel) [reply tests/pytorch/testproblems/test_noise_free_quadratic.py:58]: True, but I don't know how to do this either

**Comment by ltatzel** [comment]
@f-dangel, I tried to incorporate all your comments. Maybe double-check the unresolved conversations before merging this.
